x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BasicTokenFilterTests.java

```diff
@@ -0,0 +1,134 @@
+import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;
+import org.apache.lucene.analysis.tokenattributes.OffsetAttribute;
+
+import java.io.IOException;
+import java.io.Reader;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Some test cases taken from
+ * https://github.com/huggingface/transformers/blob/ba8c4d0ac04acfcdbdeaed954f698d6d5ec3e532/tests/test_tokenization_bert.py
+ */
+public class BasicTokenFilterTests extends BaseTokenStreamTestCase {
+
+    public void testNeverSplit_GivenNoLowerCase() throws IOException {
+        Analyzer analyzer = basicAnalyzerFromSettings(false, false, List.of("[UNK]"));
+        assertAnalyzesToNoCharFilter(analyzer, "1 (return) [ Patois ", new String[] { "1", "(", "return", ")", "[", "Patois" });
+        assertAnalyzesToNoCharFilter(analyzer, "Hello [UNK].", new String[] { "Hello", "[UNK]", "." });
+        assertAnalyzesToNoCharFilter(analyzer, "Hello-[UNK]", new String[] { "Hello", "-", "[UNK]" });
+        assertAnalyzesToNoCharFilter(
+            analyzer,
+            " \tHeLLo!how  \n Are yoU? [UNK]",
+            new String[] { "HeLLo", "!", "how", "Are", "yoU", "?", "[UNK]" }
+        );
+        assertAnalyzesToNoCharFilter(analyzer, "Hello [UNK]?", new String[] { "Hello", "[UNK]", "?" });
+        assertAnalyzesToNoCharFilter(analyzer, "Hello [UNK]!!", new String[] { "Hello", "[UNK]", "!", "!" });
+        assertAnalyzesToNoCharFilter(analyzer, "Hello~[UNK][UNK]", new String[] { "Hello", "~", "[UNK]", "[UNK]" });
+        assertAnalyzesToNoCharFilter(analyzer, "Hello-[unk]", new String[] { "Hello", "-", "[", "unk", "]" });
+    }
+
+    public void testNeverSplit_GivenLowerCase() throws IOException {
+        Analyzer analyzer = basicAnalyzerFromSettings(false, false, List.of("[UNK]"));
+        assertAnalyzesToNoCharFilter(
+            analyzer,
+            " \tHeLLo!how  \n Are yoU? [UNK]",
+            new String[] { "HeLLo", "!", "how", "Are", "yoU", "?", "[UNK]" }
+        );
+        assertAnalyzesToNoCharFilter(analyzer, "Hello [UNK].", new String[] { "Hello", "[UNK]", "." });
+        assertAnalyzesToNoCharFilter(analyzer, "Hello [UNK]?", new String[] { "Hello", "[UNK]", "?" });
+        assertAnalyzesToNoCharFilter(analyzer, "Hello [UNK]!!", new String[] { "Hello", "[UNK]", "!", "!" });
+        assertAnalyzesToNoCharFilter(analyzer, "Hello-[UNK]", new String[] { "Hello", "-", "[UNK]" });
+        assertAnalyzesToNoCharFilter(analyzer, "Hello~[UNK][UNK]", new String[] { "Hello", "~", "[UNK]", "[UNK]" });
+        assertAnalyzesToNoCharFilter(analyzer, "Hello-[unk]", new String[] { "Hello", "-", "[", "unk", "]" });
+    }
+
+    public void testSplitCJK() throws Exception {
+        Analyzer analyzer = basicAnalyzerFromSettings(true, false, List.of("[UNK]"));
+        assertAnalyzesToNoCharFilter(analyzer, "hello ah\u535A\u63A8zz", new String[] { "hello", "ah", "\u535A", "\u63A8", "zz" });
```

Comment:
nit: what happens for surrogate pairs? are they getting split or are they 1 token?